cmd/backup-manager/app/util/generic.go

```diff
@@ -76,21 +76,56 @@ func (bo *GenericOptions) GetDSN(enabledTLSClient bool) (string, error) {
 }
 
 func (bo *GenericOptions) GetTikvGCLifeTime(ctx context.Context, db *sql.DB) (string, error) {
-	var tikvGCTime string
-	sql := fmt.Sprintf("select variable_value from %s where variable_name= ?", constants.TidbMetaTable)
-	row := db.QueryRowContext(ctx, sql, constants.TikvGCVariable)
-	err := row.Scan(&tikvGCTime)
+	var tikvGCTime, sql, gcVariable string
+	has, err := bo.HasTiDBGCLifeTime(ctx, db)
+	if err != nil {
+		return "", err
+	}
+	if has {
+		gcVariable = constants.TikvGCVariable
+		sql = fmt.Sprintln("select @@?")
+	} else {
+		gcVariable = constants.LegacyTikvGCVariable
+		sql = fmt.Sprintf("select variable_value from %s where variable_name= ?", constants.TidbMetaTable)
+	}
+	row := db.QueryRowContext(ctx, sql, gcVariable)
+	err = row.Scan(&tikvGCTime)
 	if err != nil {
 		return tikvGCTime, fmt.Errorf("query cluster %s %s failed, sql: %s, err: %v", bo, constants.TikvGCVariable, sql, err)
 	}
 	return tikvGCTime, nil
 }
 
 func (bo *GenericOptions) SetTikvGCLifeTime(ctx context.Context, db *sql.DB, gcTime string) error {
-	sql := fmt.Sprintf("update %s set variable_value = ? where variable_name = ?", constants.TidbMetaTable)
-	_, err := db.ExecContext(ctx, sql, gcTime, constants.TikvGCVariable)
+	var sql, gcVariable string
+	has, err := bo.HasTiDBGCLifeTime(ctx, db)
+	if err != nil {
+		return err
+	}
+	if has {
+		gcVariable = constants.TikvGCVariable
+		sql = fmt.Sprintln("SET GLOBAL ? = ?")
+	} else {
+		gcVariable = constants.LegacyTikvGCVariable
+		sql = fmt.Sprintf("update %s set variable_value = ? where variable_name = ?", constants.TidbMetaTable)
+	}
+	_, err = db.ExecContext(ctx, sql, gcTime, gcVariable)
```

Comment:
for the `has` branch the param order should be "gcVariable, gcTime" instead of "gcTime, gcVariable"

I suggest you format the final SQL directly

Reply:
update